### PR TITLE
Updates internals and extensions chapters

### PR DESCRIPTION
**Pull request by djnavarro**
A first pass at smoothing the ggplot2 internals and extensions chapters: 

- For the internals chapter the changes are all in the ggproto section and are focused on extending the "Person" example (e.g., "Royalty" and "Police" both subclass "Person"). 
- I've made more extensive changes in the extensions chapter, providing small examples of each kind of ggplot2 extension (some of which are shamelessly stolen from vignettes, old issues threads, and in one case from ggforce) to illustrate the process. What I'm trying to do in each case is give novice developers a little more scaffolding, so that there aren't quite so many unfamiliar concepts once they hit the full "springs" case study in the final chapter (which I've not yet touched except in trivial ways).... but being a novice ggplot2 developer myself I'm not sure how well I've done it or what important points I've missed, and I think this is at the point where it needs suggestions from experts! 😀 

The other big PR in the pipeline at the moment is the scales toolbox edits (#260), and I've checked that this doesn't conflict with that one. I was thinking that since I've already dropped #260 on @hadley, and this is focused on chapters @thomasp85 has worked on, maybe Thomas could take a look at this and make suggestions?

### CODE REVIEW

File: internals.Rmd [hunk 1]
```diff
@@ -187,88 +187,227 @@ grid::grid.newpage()
...
+## Introducing ggproto {#ggproto}
+
+Sections \@ref(plot-method) to \@ref(ggplotgtable) focus on the sequence of events involved in building a ggplot, but are intentionally vague as to what kind of programming objects perform this work. 
+All ggplot2 objects are built using the **ggproto** system for object oriented programming, and is unusual in that it is used only by ggplot2. This something of a historical accident: ggplot2 originally used proto [@proto] for object oriented programming, which became a problem once the need for an official extension mechanism arose due to the limitations of the proto system. Attempts to switch ggplot2 to other systems such as R6 [@R6] proved difficult, and creating an object oriented system specific to the needs of ggplot2 turned out to be the least bad solution.
```
Comment (thomasp85) [inline internals.Rmd:193]:
```suggestion
All ggplot2 objects are built using the **ggproto** system for object oriented programming, and is unusual in that it is used only by ggplot2. This is something of a historical accident: ggplot2 originally used proto [@proto] for object oriented programming, which became a problem once the need for an official extension mechanism arose due to the limitations of the proto system. Attempts to switch ggplot2 to other systems such as R6 [@R6] proved difficult, and creating an object oriented system specific to the needs of ggplot2 turned out to be the least bad solution. 
```

File: internals.Rmd [hunk 2]
```diff
@@ -187,88 +187,227 @@ grid::grid.newpage()
...
+All ggplot2 objects are built using the **ggproto** system for object oriented programming, and is unusual in that it is used only by ggplot2. This something of a historical accident: ggplot2 originally used proto [@proto] for object oriented programming, which became a problem once the need for an official extension mechanism arose due to the limitations of the proto system. Attempts to switch ggplot2 to other systems such as R6 [@R6] proved difficult, and creating an object oriented system specific to the needs of ggplot2 turned out to be the least bad solution. 
+
+Understanding the ggproto object oriented programming system is important if you wish to write ggplot2 extensions. We will encounter ggproto objects as they are used by ggplot2 in Chapters \@ref(extensions) and \@ref(spring1). Like the better-known R6 system, ggproto uses reference semantics, allowing inheritance and access to methods from parent classes, and is accompanied by a set of design principles that, while not enforced by ggproto, is essential to understanding how the system is used in ggplot2. To illustrate these concepts, this section introduces the core mechanics of ggproto in a simplified form.
```
Comment (thomasp85) [inline internals.Rmd:195]:
```suggestion
Understanding the ggproto object oriented programming system is important if you wish to write ggplot2 extensions. We will encounter ggproto objects as they are used by ggplot2 in Chapters \@ref(extensions) and \@ref(spring1). Like the better-known R6 system, ggproto uses reference semantics and allows inheritance and access to methods from parent classes. It is accompanied by a set of design principles that, while not enforced by ggproto, is essential to understanding how the system is used in ggplot2. To illustrate these concepts, this section introduces the core mechanics of ggproto in a simplified form.
```

**Review by thomasp85 — COMMENTED** [review]
This looks good and most of my comments are superficial. You've done a great job with the additional examples

**Comment by djnavarro** [comment]
I've managed to update this PR to address the original comments, resolved the merge conflicts that were my own fault for leaving it dormant too long, and updated it to accommodate the new linewidth aesthetic. I'm still not happy with it, but it's an improvement on the current state so I'll merge this as-is and seek to fix shortcomings in a new PR

**Comment by hadley** [comment]
Thanks!

**Pull request merged**